<?xml version="1.0"?>
<launch>
    <arg name="lidar_frame_id" default="velodyne" />
    <arg name="path_to_bag" default="$(env HOME)/Downloads/example.bag" />
    <arg name="points_topic" default="points" />
    <!-- node manager arguments -->
    <arg name="node_name" default="floam" />
    <arg name="manager_name" value="$(arg node_name)_manager" />
    <arg name="global_manager_name" value="/$(arg manager_name)" />
    <arg name="new_manager" default="true" />

    <!-- needed for rosbag replay processing -->
    <param name="/use_sim_time" value="true" />

    <!-- your bag file here -->
    <node pkg="rosbag" type="play" name="rosbag_play" args="--clock $(arg path_to_bag)"/> 

    <!-- Nodelet Manager -->
    <node pkg="nodelet" type="nodelet" name="$(arg manager_name)" args="manager" output="screen" if="$(arg new_manager)" />

    <!-- floam lidar node -->
    <node pkg="nodelet" type="nodelet" name="$(arg node_name)_lidar_node"
	    args="load floam/lidar/LidarProcessingNode $(arg global_manager_name)" output="screen">
        <param name="points_topic" value="$(arg points_topic)" />  <!-- dont forget leading `/` for argument-->
        <param name="scan_lines" value="32" />
        <param name="period" value="0.1" />

        <param name="vertical_angle" type="double" value="2.0" />
        <param name="map_resolution" type="double" value="0.2" />
        <param name="max_dis" type="double" value="90.0" />
        <param name="min_dis" type="double" value="0.5" />
    </node>
	    
    <!-- floam odom node -->
    <node pkg="nodelet" type="nodelet" name="$(arg node_name)_odom_node"
      args="load floam/odom/OdomEstimationNode $(arg global_manager_name)" output="screen"/>

    <!-- floam mapping node -->
    <node pkg="nodelet" type="nodelet" name="$(arg node_name)_mapping_node"
      args="load floam/lidar/LidarMappingNode $(arg global_manager_name)" output="screen"/>
    
    <node pkg="tf" type="static_transform_publisher" name="map2baselink_tf"  args="0 0 0 0 0 0 /map /base_link 10" />
    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find floam)/rviz/floam.rviz" />
    </group>

    
  	<node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="gt" >
        <param name="/target_frame_name" value="map" />
        <param name="/source_frame_name" value="$(arg lidar_frame_id)" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="base_link" >
        <param name="/target_frame_name" value="map" />
        <param name="/source_frame_name" value="base_link" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>

</launch>
